src/OpenTelemetry.Instrumentation.Http/HttpClientMetricsInstrumentationOptions.cs

```diff
@@ -0,0 +1,88 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Diagnostics;
+using System.Net;
+#if NETFRAMEWORK
+using System.Net.Http;
+#endif
+using System.Runtime.CompilerServices;
+using OpenTelemetry.Instrumentation.Http.Implementation;
+
+namespace OpenTelemetry.Instrumentation.Http;
+
+/// <summary>
+/// Options for HttpClient metrics instrumentation.
+/// </summary>
+public class HttpClientMetricsInstrumentationOptions
+{
+    /// <summary>
+    /// Delegate for enrichment of recorded metric with additional tags.
+    /// </summary>
+    /// <param name="response"><see cref="HttpResponseMessage"/>: the HttpResponseMessage object.</param>
+    /// <param name="tags"><see cref="TagList"/>: List of current tags. You can add additional tags to this list. </param>
+    public delegate void HttpClientMetricEnrichmentFunc(HttpResponseMessage response, ref TagList tags);
+
+    /// <summary>
+    /// Gets or sets a filter function that determines whether or not to
+    /// collect telemetry on a per request basis.
+    /// </summary>
+    /// <remarks>
+    /// <para><b>FilterHttpRequestMessage is only executed on .NET and .NET
+    /// Core runtimes. <see cref="HttpClient"/> and <see
+    /// cref="HttpWebRequest"/> on .NET and .NET Core are both implemented
+    /// using <see cref="HttpRequestMessage"/>.</b></para>
+    /// Notes:
+    /// <list type="bullet">
+    /// <item>The return value for the filter function is interpreted as:
+    /// <list type="bullet">
+    /// <item>If filter returns <see langword="true" />, the request is
+    /// collected.</item>
+    /// <item>If filter returns <see langword="false" /> or throws an
+    /// exception the request is NOT collected.</item>
+    /// </list></item>
+    /// </list>
+    /// </remarks>
+    public Func<HttpRequestMessage, bool> FilterHttpRequestMessage { get; set; }
+
+    /// <summary>
+    /// Gets or sets an function to enrich a recorded metric with additional custom tags.
+    /// </summary>
+    public HttpClientMetricEnrichmentFunc Enrich { get; set; }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal bool EventFilterHttpRequestMessage(string activityName, object arg1)
+    {
+        try
+        {
+            return
+                this.FilterHttpRequestMessage == null ||
+                !TryParseHttpRequestMessage(activityName, arg1, out HttpRequestMessage requestMessage) ||
+                this.FilterHttpRequestMessage(requestMessage);
+        }
+        catch (Exception ex)
+        {
+            HttpInstrumentationEventSource.Log.RequestFilterException(ex);
+            return false;
+        }
```

Comment:
nit: Might be a bit faster to bypass the try/catch if filter delegate is null eg:

```csharp
        var filter = this.FilterHttpRequestMessage;
        if (filter == null) 
        {
              return true;
        }

        try
        {
            return !TryParseHttpRequestMessage(activityName, arg1, out HttpRequestMessage requestMessage)
                || filter(requestMessage);
        }
        catch (Exception ex)
        {
            HttpInstrumentationEventSource.Log.RequestFilterException(ex);
            return false;
        }
```

Reply:
The PR pending for this review comment or it's because we have to discuss the whole topic?

@CodeBlanch @reyang